source/Halibut.Tests/BadCertificatesTests.cs

```diff
@@ -29,6 +29,9 @@ public async Task SucceedsWhenPollingServicePresentsWrongCertificate_ButServiceI
             var clientTrustProvider = new DefaultTrustProvider();
             var unauthorizedThumbprint = "";
             var firstCall = true;
+            
+            var unauthorizedClientHasConnected = new TaskCompletionSource<bool>();
+            CancellationToken.Register(() => unauthorizedClientHasConnected.TrySetCanceled()); // backup to fail the test in case it never connects
```

Comment:
nit: could be unregistered

Reply:
There's no value in unregistering things in the same object graph.

Put another way:

On the "cancelled" code path:
When the CancellationToken fires, the test will tear down, the class/variable/etc will go out of scope and the garbage collector will eat the whole lot in one go.

On the "normal" code path:
TearDown in BaseTest will cancel and dispose the CancellationTokenSource, which will clean up all this stuff; the garbage collector will eat the whole lot there too.